Add Tutorial Running Kubelet in Standalone Mode (#48094)
kubernetes/website: content/en/docs/tutorials/configuration/kubelet-standalone.md

Reviewer: On reflection, I think this merits a new section (not Configuration). Maybe **Cluster Management**?

Within this new section, I'd give it weight 10. The Cluster Management section should come after Security, so give it a fairly high weight number (eg 60).

Author: Let me try that in another commit. Thanks

Author: Done

---
title: Running Kubelet in Standalone Mode
content_type: tutorial
weight: 5
card:
  name: tutorials
  weight: 10
---

<!-- overview -->

> This tutorial shows you how to run a standalone kubelet instance.

Author: Done

## {{% heading "objectives" %}}

* Install and configure `containerd`, `cni-plugins`, and `kubelet` on a Linux system.

> * Run `containerd` as a `systemd` service and run `kubelet` on the command line.

Author: Done

* Launch a Pod running `nginx` that listens to requests on TCP port 80 on the Pod's IP address.
* View how the various components work together to run the Pod and analyze their outputs and logs.

## {{% heading "prerequisites" %}}

- Admin (root) access to a Linux system (virtual, cloud, or physical instance) running `systemd`.

> - Access to the Internet to download the various components needed for the lab.

Author: Done

- Required CLI tools: `curl, tar`.

> - Useful CLI tools: `ps, ip, netstat, iptables, tree, jq`.

Author: Done

EOF
```

- Install:

```shell
tar Cxzvf /usr/local/ containerd-1.7.22-linux-amd64.tar.gz
mv containerd.service /etc/systemd/system/
systemctl enable containerd.service
systemctl start containerd.service
```

- Quick test:

```shell
systemctl status containerd.service
```

### Kubelet

[latest stable release](https://cdn.dl.k8s.io/release/stable.txt)

- Download:

```shell
curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubelet"
```

- Configure:
[docs](/docs/tasks/administer-cluster/kubelet-config-file/)

```shell
mkdir -p /etc/kubernetes/kubelet.conf.d

cat << EOF > /etc/kubernetes/kubelet.conf.d/kubelet.yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
authentication:
  webhook:
    enabled: false
authorization:
  mode: AlwaysAllow
enableServer: false
logging:
  format: text
  sanitization: false
readOnlyPort: 10250
podCIDR: 10.241.1.0/24
staticPodPath: /etc/kubernetes/manifests
EOF
``` 

- Install:

```shell
chmod +x kubelet
mv kubelet /usr/bin/
```

- Quick test:

```shell
kubelet --version
```

The output is similar to:

```
Kubernetes v1.31.1
```

- Run `kubelet`:

```shell

> kubelet --config=/etc/kubernetes/kubelet.conf.d/kubelet.yaml

Author: Done

```

The output is similar to:

```
    inet 10.241.1.1/32 scope global veth85d4e050
    inet6 fe80::f04e:1ff:fefd:9464/64 scope link
```

That's your Pod's IP address.

Query the Kubelet's API at `http://<IP>:<Port>/pods`.

For this example, it's `http://10.241.1.1:10250/pods`

```shell
curl http://10.241.1.1:10250/pods
```

Tip: use `jq` to "prettify" the JSON output:

```shell
curl http://10.241.1.1:10250/pods | jq '.'
```

To obtain the IP address of the `nginx` Pod:

```shell
curl http://10.241.1.1:10250/pods | jq '.items[].status.podIP'
```

The output is similar to:

```
"10.241.1.11"
```

Connect to the `nginx` server on `http://<IP>:<Port>` (port 80 is the default), in this case:

```shell
curl http://10.241.1.11
```

The output is similar to:

```
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
...
```

- Some directories of interest:

```
/var/lib/cni
/var/lib/containerd
/var/lib/kubelet

/var/log/containers
/var/log/pods
```

## Clean up

Now you can clean up the resources you created for this lab or leave them for further
investigation.

Tip: You can gather all the CLI commands and put them in a shell script that you can

> run if you want to replicate the lab environment.

Reviewer: We should explain _how_ to clean up.

Author: Done

---
title: Running Kubelet in Standalone Mode
content_type: tutorial
weight: 5
card:
  name: tutorials
  weight: 10
---

<!-- overview -->

This tutorial shows you how to run a standalone Kubelet instance.

In standalone mode the Kubelet allows you to manage containers using pod manifests, which brings the benefits of running tightly coupled application as a single unit while leveraging a subset of advanced features including init containers, CNI networking, etc., without the need for the entire Kubernetes feature set.

## {{% heading "objectives" %}}

* Install and configure `containerd`, `cni-plugins`, and `kubelet` on a Linux system.
* Run `containerd` and `kubelet` as `systemd` services.
* Launch a Pod running `nginx` that listens to requests on TCP port 80 on the Pod's IP address.

## {{% heading "prerequisites" %}}

> 

Reviewer: ```suggestion

You need a computer with an AMD64 or compatible processor. This tutorial does not provide
instructions for other architectures.
You also need:

```
(if we could cover, eg, ARM64 as well, that would be nice)

Author: I don't have access to an ARM64 system to test but I added the `arm64` download URLs for all the components and duplicated the download and the install commands in tabs format for both `amd64` and `arm64`.

Reviewer: BTW, you can use (eg) qemu to test locally without buying new equipment. I think the testing you said you did is enough though.

- Admin (root) access to a Linux system (virtual, cloud, or physical instance) running `systemd`.
- Access to the Internet to download the various components needed for the tutorial.
- Required CLI tools: `curl`, `tar`.
- Suggested CLI tools: `jq`.

<!-- lessoncontent -->

## Download, install, and configure the components

{{< note >}}
Download the latest available versions of the required packages.
The relative path to the downloaded software is assumed to be the `sources` directory.
Run all the commands as the `root` user.
{{< /note >}}

```shell
mkdir sources
cd sources/
```

### CNI plugins

> Download: [[releases]](https://github.com/containernetworking/plugins/releases)

Reviewer: ```suggestion
This tutorial suggests downloading sample network plugins. Visit [releases](https://github.com/containernetworking/plugins/releases)
(external link) for more information.
```
?

Author: Done

```shell
curl -LO https://github.com/containernetworking/plugins/releases/download/v1.5.1/cni-plugins-linux-amd64-v1.5.1.tgz
```

Install:

```shell
mkdir -p /opt/cni/bin
tar Cxvf /opt/cni/bin/ cni-plugins-linux-amd64-v1.5.1.tgz
```

Quick test:

```shell
/opt/cni/bin/host-local --version
```

The output is similar to:

```
CNI host-local plugin v1.5.1
CNI protocol versions supported: 0.1.0, 0.2.0, 0.3.0, 0.3.1, 0.4.0, 1.0.0
```

Configure:

```shell
mkdir -p /etc/cni

cat << EOF > /etc/cni/cni.template
{
  "name": "containerd",
  "cniVersion": "1.0.0",
  "plugins": [{
    "type": "ptp",
    "ipMasq": true,
    "ipam": {
      "type": "host-local",
      "subnet": "{{.PodCIDR}}",
      "routes": [
        {"dst": "0.0.0.0/0"}
      ]
    }
  },{
    "type": "portmap",
    "capabilities": {
      "portMappings": true
    }
  }]
}
EOF
```

### containerd

{{% thirdparty-content %}}

Enable IPv4 packet forwarding:

```shell
cat <<EOF > /etc/sysctl.d/k8s.conf
net.ipv4.ip_forward = 1
EOF

sysctl --system
```

The output is similar to:

```
...
* Applying /etc/sysctl.d/k8s.conf ...
net.ipv4.ip_forward = 1
* Applying /etc/sysctl.conf ...
```

> Download: [[releases]](https://github.com/containerd/containerd/releases)

Reviewer: ```suggestion
Download containerd (a container runtime that implements the Kubernetes 
{{<glossary_tooltip term_id="cri" text="CRI">}}):
```

Author: I added the network plugins and container runtime tooltips to the prerequisites section. See if you like it.

```shell
curl -LO https://github.com/containerd/containerd/releases/download/v1.7.22/containerd-1.7.22-linux-amd64.tar.gz
curl -LO https://raw.githubusercontent.com/containerd/containerd/main/containerd.service 
```

Configure:

```shell
mkdir -p /etc/containerd

cat << EOF > /etc/containerd/config.toml
version = 2

[plugins."io.containerd.grpc.v1.cri".cni]

    # binaries
    bin_dir = "/opt/cni/bin"

    # configure pod networking
    conf_template = "/etc/cni/cni.template"
EOF
```

Install:

```shell
tar Cxzvf /usr/local/ containerd-1.7.22-linux-amd64.tar.gz
mv containerd.service /etc/systemd/system/

> systemctl enable containerd.service
> systemctl start containerd.service

Reviewer: ```suggestion
systemctl enable --now containerd.service
```

Author: Done

```

Quick test:

```shell
systemctl status containerd.service
```

Detailed service check:

```shell
journalctl -u containerd
```

### Kubelet

> Download: [[latest stable release]](https://cdn.dl.k8s.io/release/stable.txt)

Reviewer: I'd rather not encourage bookmarking this URL.
```suggestion
Download the latest stable release (to see versions available, visit the [downloads](/releases/download/) page.
```

Author: Done

```shell
curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubelet"
```

Configure:

```shell
mkdir -p /etc/kubernetes

cat << EOF > /etc/kubernetes/kubelet.yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
authentication:
  webhook:
    enabled: false
authorization:
  mode: AlwaysAllow
enableServer: false
logging:
  format: text
  sanitization: false
readOnlyPort: 10250

> podCIDR: 10.241.1.0/24

Author: I made it up. I've added a warning note and let the user decide which network they want to use that doesn't collide with their existing networks

staticPodPath: /etc/kubernetes/manifests
EOF
```

Install:

```shell
chmod +x kubelet
mv kubelet /usr/bin/

cat <<EOF > /etc/systemd/system/kubelet.service
[Unit]
Description=Kubelet

[Service]
ExecStart=/usr/bin/kubelet \
 --config=/etc/kubernetes/kubelet.yaml
Restart=always

[Install]
WantedBy=multi-user.target
EOF

> systemctl enable kubelet.service
> systemctl start kubelet.service

Reviewer: ```suggestion
systemctl daemon-reload
systemctl enable --now kubelet.service
```

Author: Done

```

Quick test:

```shell
systemctl status kubelet.service
```

Detailed service check:

```shell
journalctl -u kubelet
```

## Run a static Pod in the Kubelet

Configure:

```shell
cat <<EOF > static-web.yaml
apiVersion: v1
kind: Pod
metadata:
  name: static-web
  labels:
    role: myrole
spec:
  containers:
    - name: web
      image: nginx
      ports:
        - name: web
          containerPort: 80
          protocol: TCP
EOF
```

Deploy:

```shell
mkdir -p /etc/kubernetes/manifests
```

Move the `static-web.yaml` manifest file to the `/etc/kubernetes/manifests` directory.

```shell

> mv static-web-yaml /etc/kubernetes/manifests/

Reviewer 2: ```suggestion
mv static-web.yaml /etc/kubernetes/manifests/
```

Author: Done. Good catch :)

```

Use the output to retrieve the IP address of the `veth` interface:

```shell
ip addr show veth85d4e050 | grep inet
```

The output is similar to:

```
    inet 10.241.1.1/32 scope global veth85d4e050
    inet6 fe80::f04e:1ff:fefd:9464/64 scope link
```

Using the IP address of the veth interface, query the Kubelet's API endpoint at `http://10.241.1.1:10250/pods`:

```shell
curl http://10.241.1.1:10250/pods
```

Tip: Use `jq` to "prettify" the JSON output:

```shell
curl http://10.241.1.1:10250/pods | jq '.'
```

To obtain the IP address of the `nginx` Pod:

```shell
curl http://10.241.1.1:10250/pods | jq '.items[].status.podIP'
```

The output is similar to:

```
"10.241.1.11"
```

Connect to the `nginx` server on `http://<IP>:<Port>` (port 80 is the default), in this case:

```shell
curl http://10.241.1.11
```

The output is similar to:

```
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
...
```

Directories for monitoring and troubleshooting:

```
/var/lib/cni
/var/lib/containerd
/var/lib/kubelet

/var/log/containers
/var/log/pods
```

## Clean up

Kubelet:

```shell

> systemctl stop kubelet
> systemctl disable kubelet

Reviewer: ```suggestion
systemctl disable --now kubelet.service
```

Author: Done

rm /usr/bin/kubelet
rm -rf /etc/kubernetes
rm -rf /var/lib/kubelet
rm -rf /var/log/containers
rm -rf /var/log/pods
```

Containerd:

```shell

> systemctl stop containerd
> systemctl disable containerd

Reviewer: ```suggestion
systemctl disable --now containerd.service
```
Is there a `.socket` unit to disable as well?

Author: Done

rm /usr/local/bin/containerd
rm -rf /etc/containerd
rm -rf /var/lib/containerd
```

> cni-plugins:

Reviewer: ```suggestion
Remove network plugins:
```

Author: Renamed cni-plugins to Network Plugins but "Remove" in this context is redundant, as it is the cleanup section and the `rm` command is self-explanatory

```shell
rm -rf /opt/cni
rm -rf /etc/cni
rm -rf /var/lib/cni
```

> 

Reviewer: What about the systemd unit files; should we clean those up too?

Author: Done 

I would not use my personal or work computer to install system software as `root`, it's too risky. I would use a disposable system such as a virtual machine or a cloud instance for this tutorial.

## {{% heading "objectives" %}}

* Install and configure `containerd`, `cni-plugins`, and `kubelet` on a Linux system.
* Run `containerd` and `kubelet` as `systemd` services.
* Launch a Pod running `nginx` that listens to requests on TCP port 80 on the Pod's IP address.

## {{% heading "prerequisites" %}}

- Admin (root) access to a Linux system (virtual, cloud, or physical instance) running `systemd`.
- Access to the Internet to download the various components needed for the tutorial.
- Required CLI tools: `curl`, `tar`.
- Suggested CLI tools: `jq`.

<!-- lessoncontent -->

## Download, install, and configure the components

{{< note >}}
Download the latest available versions of the required packages.
The relative path to the downloaded software is assumed to be the `sources` directory.
Run all the commands as the `root` user.
{{< /note >}}

```shell
mkdir sources
cd sources/
```

### CNI plugins

> - Download:
> 
> [releases](https://github.com/containernetworking/plugins/releases)
> 
> ```shell
> curl -LO https://github.com/containernetworking/plugins/releases/download/v1.5.1/cni-plugins-linux-amd64-v1.5.1.tgz
> ```
> 

Reviewer 3: Please indent the list item contents properly. Same below.

````suggestion
- Download:

  [releases](https://github.com/containernetworking/plugins/releases)

  ```shell
  curl -LO https://github.com/containernetworking/plugins/releases/download/v1.5.1/cni-plugins-linux-amd64-v1.5.1.tgz
  ```

````

Author: Actually, I realize it was a mistake to use list items to highlight those entries, I rather use plain text, like in other tutorial pages here. Thanks.

---
title: Running Kubelet in Standalone Mode
content_type: tutorial
weight: 5

> card:
>   name: tutorials
>   weight: 10

Reviewer: Just realized; I would drop this
```suggestion
```

Author: Done

{{% thirdparty-content %}}

{{< note >}}
Download the latest available versions of the required packages.
The relative path to the downloaded software is assumed to be the `sources` directory.
{{< /note >}}

{{< note >}}
Run all the commands as the `root` user.
{{< /note >}}

```shell
mkdir sources
cd sources/
```

### Network Plugins

This tutorial suggests downloading CNI network plugins.
Visit [releases](https://github.com/containernetworking/plugins/releases) (external link).

{{< tabs name="download_cni_plugins" >}}
{{< tab name="x86-64" codelang="bash" >}}
curl -LO https://github.com/containernetworking/plugins/releases/download/v1.5.1/cni-plugins-linux-amd64-v1.5.1.tgz
{{< /tab >}}
{{< tab name="ARM64" codelang="bash" >}}
curl -LO https://github.com/containernetworking/plugins/releases/download/v1.5.1/cni-plugins-linux-arm64-v1.5.1.tgz
{{< /tab >}}
{{< /tabs >}}

<!--
```shell
curl -LO https://github.com/containernetworking/plugins/releases/download/v1.5.1/cni-plugins-linux-amd64-v1.5.1.tgz
```
-->

Install:

```shell
mkdir -p /opt/cni/bin
```

{{< tabs name="install_cni_plugins" >}}
{{< tab name="x86-64" codelang="bash" >}}

> tar Cxvf /opt/cni/bin/ cni-plugins-linux-amd64-v1.5.1.tgz

Reviewer: This is likely to need `sudo` to actually work.

Author: Almost all the commands in this tutorial require `root` access, that's why I put a big note at the beginning saying so. It assumes you log on as or become `root`. But I will prefix them all with `sudo` anyway, just to be sure.

{{< /tab >}}
{{< tab name="ARM64" codelang="bash" >}}
tar Cxvf /opt/cni/bin/ cni-plugins-linux-arm64-v1.5.1.tgz
{{< /tab >}}
{{< /tabs >}}

<!--
```shell
mkdir -p /opt/cni/bin
tar Cxvf /opt/cni/bin/ cni-plugins-linux-amd64-v1.5.1.tgz
```
-->

Quick test:

```shell
/opt/cni/bin/host-local --version
```

The output is similar to:

```
CNI host-local plugin v1.5.1
CNI protocol versions supported: 0.1.0, 0.2.0, 0.3.0, 0.3.1, 0.4.0, 1.0.0
```

Configure:

```shell

> mkdir -p /etc/cni

Reviewer: This is likely to need `sudo` to actually work.

Author: Done

cat << EOF > /etc/cni/cni.template
{
  "name": "containerd",
  "cniVersion": "1.0.0",
  "plugins": [{
    "type": "ptp",
    "ipMasq": true,
    "ipam": {
      "type": "host-local",
      "subnet": "{{.PodCIDR}}",
      "routes": [
        {"dst": "0.0.0.0/0"}
      ]
    }
  },{
    "type": "portmap",
    "capabilities": {
      "portMappings": true
    }
  }]
}
EOF
```

> ### Container Runtime

Reviewer: ```suggestion
### Container runtime
```

Author: Done

> This tutorial suggests downloading the containerd container runtime.

Reviewer: Completely optional, but might help with learning:
```suggestion
This tutorial suggests downloading the containerd container runtime. When you
set up containerd, it needs a lower-level container runtime such as `runc`. These steps
also cover the setup steps you need to ensure that `runc` works.
```
(is that right? Does `runc` get set up - I assume it does, but maybe we should check)

Author: You're right. I overlooked `runc` because I forgot I've installed it as a .deb pkg. But I've added instructions for downloading and installing it. No additional configuration is required, just the binary, looks like `containerd` handles it.

Visit [releases](https://github.com/containerd/containerd/releases) (external link).

Enable IPv4 packet forwarding:

```shell
cat <<EOF > /etc/sysctl.d/k8s.conf
net.ipv4.ip_forward = 1
EOF

sysctl --system
```
The output is similar to:

```
...
* Applying /etc/sysctl.d/k8s.conf ...
net.ipv4.ip_forward = 1
* Applying /etc/sysctl.conf ...
```

Download:

{{< tabs name="download_containerd" >}}
{{< tab name="x86-64" codelang="bash" >}}
curl -LO https://github.com/containerd/containerd/releases/download/v1.7.22/containerd-1.7.22-linux-amd64.tar.gz
{{< /tab >}}
{{< tab name="ARM64" codelang="bash" >}}
curl -LO https://github.com/containerd/containerd/releases/download/v1.7.22/containerd-1.7.22-linux-arm64.tar.gz
{{< /tab >}}
{{< /tabs >}}

```shell
curl -LO https://raw.githubusercontent.com/containerd/containerd/main/containerd.service
```

<!--
```shell
curl -LO https://github.com/containerd/containerd/releases/download/v1.7.22/containerd-1.7.22-linux-amd64.tar.gz
curl -LO https://raw.githubusercontent.com/containerd/containerd/main/containerd.service 
```
-->

> Configure:

Reviewer: Is this true? **Adding text here is optional.**
```suggestion
The CRI support in containerd is already active, but you need to configure a (node local)
pod network (read
[The Kubernetes Network Model](/docs/concepts/services-networking/#the-kubernetes-network-model))
 for more context.

To set up a pod network:
```

Author: Not sure about this question. I didn't have to do anything else to make it work beyond what I wrote in the tutorial.

```shell
mkdir -p /etc/containerd

cat << EOF > /etc/containerd/config.toml
version = 2

[plugins."io.containerd.grpc.v1.cri".cni]

    # binaries
    bin_dir = "/opt/cni/bin"

    # configure pod networking
    conf_template = "/etc/cni/cni.template"
EOF
```

Install:

{{< tabs name="install_containerd" >}}
{{< tab name="x86-64" codelang="bash" >}}

> tar Cxzvf /usr/local/ containerd-1.7.22-linux-amd64.tar.gz

Reviewer: This command is likely to need `sudo` (or an equivalent).

Author: Done

{{< /tab >}}
{{< tab name="ARM64" codelang="bash" >}}
tar Cxzvf /usr/local/ containerd-1.7.22-linux-arm64.tar.gz
{{< /tab >}}
{{< /tabs >}}

```shell
mv containerd.service /etc/systemd/system/

> systemctl enable --now containerd.service

Reviewer: Likely to need `sudo`

Author: Done

```

<!--
```shell
tar Cxzvf /usr/local/ containerd-1.7.22-linux-amd64.tar.gz
mv containerd.service /etc/systemd/system/

systemctl enable --now containerd.service
```
-->

Quick test:

```shell
systemctl status containerd.service
```

Detailed service check:

```shell
journalctl -u containerd
```

### Kubelet

Download the [latest stable release](/releases/download/) of the Kubelet.

{{< tabs name="download_kubelet" >}}
{{< tab name="x86-64" codelang="bash" >}}
curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubelet"
{{< /tab >}}
{{< tab name="ARM64" codelang="bash" >}}
curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/arm64/kubelet"
{{< /tab >}}
{{< /tabs >}}

<!--
```shell
curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubelet"
```
-->

Configure:

{{< note >}}

> This tutorial uses the `10.241.1.0` network. Make sure you don't have any

Reviewer: ```suggestion
This tutorial uses the `10.241.1.0/24` network range (which is a private network). Make sure you
don't have any
```

Author: Done

conflicts with an existing network. If necessary, use a network that is available to
you.
{{< /note >}}

```shell

> mkdir -p /etc/kubernetes

Reviewer: Likely to need `sudo`

Author: Done

cat << EOF > /etc/kubernetes/kubelet.yaml
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
authentication:
  webhook:
    enabled: false
authorization:
  mode: AlwaysAllow
enableServer: false
logging:
  format: text
  sanitization: false
readOnlyPort: 10250
podCIDR: 10.241.1.0/24
staticPodPath: /etc/kubernetes/manifests
EOF
```

Install:

```shell
chmod +x kubelet

> mv kubelet /usr/bin/

Reviewer: - likely to need `sudo`
- for tidiness, set the owner as `root`

Author: I've changed all instances of `sudo mv...` to `sudo cp...`, it saves the extra `chown` step, and you keep the sources intact, in case you want to reuse them

> 

Reviewer: ```suggestion

# If you are logged in as an unprivileged user, you can use
# sudoedit /etc/systemd/system/kubelet.service
# instead, and paste in the contents of the unit file
```
?

Author: What's this?

cat <<EOF > /etc/systemd/system/kubelet.service
[Unit]
Description=Kubelet

[Service]
ExecStart=/usr/bin/kubelet \
 --config=/etc/kubernetes/kubelet.yaml
Restart=always

[Install]
WantedBy=multi-user.target
EOF

> 

Reviewer: ```suggestion

systemctl daemon-reload
```
On a desktop type setup this usually prompts for authentication (eg, I provide a fingerprint) so I don't feel the need to mention `sudo` here.
```

Author: Done, but I'll keep the `sudo`

systemctl enable --now kubelet.service
```

Quick test:

```shell

> systemctl status kubelet.service

Reviewer: ```suggestion
# When you run this, the output should include "Active: active (running)"
systemctl status kubelet.service
```

Author: I'm replacing `status` with `is-active`. No `sudo` and no lengthy output. Same for containerd

```

Detailed service check:

```shell
journalctl -u kubelet
```

## Run a static Pod in the Kubelet

> Configure:

Reviewer: ```suggestion
In standalone mode, you can only run [static Pods](/docs/tasks/configure-pod-container/static-pod/).
The kubelet can fetch the static Pod configuration from a local filesystem, or over a network using HTTP.

Define a manifest for example static Pod:
```

Author: Done

> 

Reviewer: nit: I would make the manifests directory first, before typing out a manifest into their current directory.

We could even consider that as part of the previous section: kubelet setup.

Author: Agree. I added the `manifests` subdir to the `mkdir` step in kubelet

```shell
cat <<EOF > static-web.yaml
apiVersion: v1
kind: Pod
metadata:
  name: static-web
  labels:
    role: myrole
spec:
  containers:
    - name: web
      image: nginx
      ports:
        - name: web
          containerPort: 80
          protocol: TCP
EOF

> ```

Reviewer: **(optional)**
This could be an example manifest using `code_sample` that we get them to fetch using `curl` and then copy to `/etc/kubernetes/manifests`.

Author: It's literally copy/pasted from the  [Static Pods](/docs/tasks/configure-pod-container/static-pod/) page

Deploy:

```shell
mkdir -p /etc/kubernetes/manifests
```

Move the `static-web.yaml` manifest file to the `/etc/kubernetes/manifests` directory.

```shell
mv static-web.yaml /etc/kubernetes/manifests/
```

### Find out information about the Kubelet and the static Pod

> The cni-plugin creates a `veth` interface, retrieve the information:

Reviewer: (nit / suggestion)
```suggestion
The Pod networking plugin creates a pair of `veth` interfaces for each Pod (one of the pair is inside
the newly made Pod, and the other is at the host level).
Retrieve information about the example Pod's network interface:
```

Author: Done

```shell

> ip link show | grep veth

Reviewer: Try:
```suggestion
ip link show type veth up
```

Author: Nope, same output, it just adds an extra line, but I truncated it to the first line, anyway

```

The output is similar to:

```
6: veth85d4e050@if2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP mode DEFAULT group default
```

> Use the output to retrieve the IP address of the `veth` interface:

Reviewer: (nit)
```suggestion
Use the output to retrieve the IPv4 address of the `veth` interface:
```

Author: Done

```shell

> ip addr show veth85d4e050 | grep inet

Reviewer: Could use something like:
```shell
ip -f inet address show dev veth12ab34bc
```

Author: Okay but the output is still very verbose, so I'll keep the `| grep inet`. I think you have a dislike for `grep` :)

Reviewer: OK; we don't need to make a change here for this to merge.

I wouldn't use `grep` on this line though; I don't think it's a good tool to teach in this context.

Or if we want to teach people how to use iproute2 a bit, I'd leave out the grep so they can see the detail in context (and, if they are a novice, they can look for the only element which is in CIDR notation).

We are helping people to _learn_ here rather than to achieve a task, so it's OK not to do every step for them.

---

Outside of a tutorial, you could use:
```shell
ip --json -f inet address show dev veth12ab34cd | jq -r '.[0] | .addr_info | .[] | ( [ .local, .prefixlen ] | join("/") )'
```
which gets you just the IP address ranges, in CIDR notation.

Author: Done. Thanks for the detailed explanation

```

The output is similar to:

```
    inet 10.241.1.1/32 scope global veth85d4e050
    inet6 fe80::f04e:1ff:fefd:9464/64 scope link
```

> Using the IP address of the veth interface, query the Kubelet's API endpoint at `http://10.241.1.1:10250/pods`:

Reviewer: Why can't we query http://localhost:10250/pods? That might be easier.

Author: You can, and it is easier. But finding out and using the real IP address seems more thorough (to me, at least)

Reviewer: It's possibly unhelpful. We can't assume that the kubelet listens on each Pod IP address; it may well not.

Author: Done

```shell
curl http://10.241.1.1:10250/pods
```

Tip: Use `jq` to "prettify" the JSON output:

```shell
curl http://10.241.1.1:10250/pods | jq '.'
```

To obtain the IP address of the `nginx` Pod:

```shell
curl http://10.241.1.1:10250/pods | jq '.items[].status.podIP'
```

The output is similar to:

```
"10.241.1.11"
```

Connect to the `nginx` server on `http://<IP>:<Port>` (port 80 is the default), in this case:

```shell
curl http://10.241.1.11
```

The output is similar to:

```
<!DOCTYPE html>
<html>
<head>
<title>Welcome to nginx!</title>
...
```

> Directories for monitoring and troubleshooting:

Reviewer: ```suggestion
## Where to look for more details

If you need to diagnose a problem getting this to work, you can look within the
following directories for monitoring and troubleshooting:
```

Author: Done

```
/var/lib/cni
/var/lib/containerd
/var/lib/kubelet

/var/log/containers
/var/log/pods
```

## Clean up

### Kubelet

```shell
systemctl disable --now kubelet.service
rm /etc/systemd/system/kubelet.service
rm /usr/bin/kubelet
rm -rf /etc/kubernetes
rm -rf /var/lib/kubelet
rm -rf /var/log/containers
rm -rf /var/log/pods
```

### Container Runtime

```shell
systemctl disable --now containerd.service
rm /etc/systemd/system/containerd.service
rm /usr/local/bin/containerd
rm -rf /etc/containerd
rm -rf /var/lib/containerd
```

### Network Plugins

```shell
rm -rf /opt/cni
rm -rf /etc/cni
rm -rf /var/lib/cni
```

## Conclusion

> This page covered the basic aspects of deploying a standalone Kubelet. You are now

Reviewer: ```suggestion
This page covered the basic aspects of deploying a kubelet in standalone mode. You are now
```

Author: Done

ready to deploy Pods and test additional functionality.

> 

Reviewer: (optional)
```suggestion
In standalone mode, the kubelet does **not** support fetching Pod configuration from the
control plane (because there is no control plane connection). You cannot use a
{{< glossary_tooltip text="ConfigMap" term_id="configmap" >}} or a
{{< glossary_tooltip text="Secret" term_id="secret" >}} to configure the containers in a static
Pod.

```

Author: Done, with some minor tweak

## {{% heading "whatsnext" %}}

* Learn more about [Network Plugins](/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/)
* Learn more about [Container Runtimes](/docs/setup/production-environment/container-runtimes/)
* Learn more about [kubelet](/docs/reference/command-line-tools-reference/kubelet/)

> * Learn more about [Static Pods](/docs/tasks/configure-pod-container/static-pod/)

Reviewer: (nit)
```suggestion
* Learn more about [static Pods](/docs/tasks/configure-pod-container/static-pod/)
```

Author: Done

Query the Kubelet's API at `http://<IP>:<Port>/pods`.

> In this example, the Pod's URL is `http://10.241.1.1:10250/pods`:

Reviewer 3: No. This URI is the kubelet's API endpoint. It is not the URL of a Pod.

Author: Done

```shell

> curl 10.241.1.1:10250/pods

Reviewer 3: Why these changes?

Author: Done
